linalg/solvers.hpp

```diff
@@ -65,24 +65,95 @@ class IterativeSolverMonitor
 /// Abstract base class for iterative solver
 class IterativeSolver : public Solver
 {
+public:
+   /** @brief Settings for the output behavior of the iterative solver. Guarantees to
+       supress all outputs by default.
+     */
+   struct PrintOptions
```

Comment:
Another way of implementing this would be with an enum/bit flags. The advantage there would be that you could combine options in one-liners, e.g. `PrintOptions::SUMMARY | PrintOptions::ITERATION_DETAILS`. This method seems fine to me though, I don't know if there are other compelling reasons one way or the other.

Reply:
Exactly my thoughts. I'd prefer on a revamp like this to introduce explicitly named bit level flags for options. The advantage is that one could predefine leveled options for nested solvers etc.

Reply:
Thanks for the suggestion @pazner . I do not have a strong opinion here, but I see that this is favorable from the user perspective. The rationale of choosing a structure over an (strong) enumeration was that it is slightly more readable in the backend: e.g. `if (print_options.iteration_details)` becomes  ` if (print_options & PrintLevel::ITERATION_DETAILS) ` . Not a huge difference or problem here.

A possible issue with using enums is namespace pollution if I take them out of the class into the MFEM namespace - if I leave them where they are the access is a bit convoluted through ` IterativeSolver::PrintLevel::SUMMARY | IterativeSolver::PrintLevel::ITERATION_DETAILS `.

Not quite understanding your point @jandrej , could you elaborate? If you want custom options for custom solvers you can also go with a custom ` PrintLevel ` inside the custom solver.

Reply:
Imagine having a nested solver GMRES(CG(AMG)) (for whatever reason). With bitflags you could set
`base_flags = SUMMARY` for GMRES and then for CG you might be interested in `base_flags | ITERATION_DETAILS`.

Reply:
Yeah, I'm a bit torn on this decision myself. I see the advantages of doing it either way (and I agree with you that `IterativeSolver::PrintLevel::SUMMARY | IterativeSolver::PrintLevel::ITERATION_DETAILS` is a bit verbose).

Reply:
I see, thanks for the elaboration @jandrej .

Regarding the verbosity of the name I do not have found anything useful yet. We can shorten the name with using directives, e.g. `using IterSolvPrintLevel = IterativeSolver::PrintLevel`, without leaking the enumerator list into the mfem namespace (currently we already leak into the IterativeSolver namespace, which is also not really optimal). Alternatively we could make this a strong enum and directly move it into the mfem namespace - but then again I do not have a good name for this construct and IterativeSolver::PrintLevel seems a bit verbose. We could probably just recommend to use `PrintLevel = IterativeSolver::PrintLevel` if one want to have a concise way to set the flags locally.